### PR TITLE
Decrease ps write latency

### PR DESCRIPTION
**Pull request by lidezhu**
### What problem does this PR solve?

Issue Number: ref #6827 

Problem Summary: We need to decrease ps write latency.

### What is changed and how it works?
Make apply to PageDirectory a pipeline so that we can batch the incoming edit when doing IO.
Before optimization
aeb252db-79fe-494a-b46a-e8ad6e6b5235
After optimization
19c406a7-932f-4ca1-834b-28e5dceac01c
The 99-latency decrease from 40ms to 12ms.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Storages/Page/V3/PageDirectory.h [hunk 1]
```diff
@@ -453,7 +463,9 @@ class PageDirectory
     std::atomic<UInt64> sequence;
 
     // Used for avoid concurrently apply edits to wal and mvcc_table_directory.
-    mutable std::shared_mutex apply_mutex;
+    mutable std::mutex apply_mutex;
+
+    std::deque<Writer *> writers;
```
[the remark this answers is not in the document]
Comment (lidezhu) [reply dbms/src/Storages/Page/V3/PageDirectory.h:481]: Added.

**Review by flowbehappy — COMMENTED** [review]
LGTM

**Review by CalvinNeo — APPROVED** [review]
lgtm

**Review by flowbehappy — APPROVED** [review]

File: dbms/src/Storages/Page/V3/PageDirectory.cpp [hunk 2]
```diff
@@ -1452,23 +1452,49 @@ void PageDirectory<Trait>::applyRefEditRecord(
...
 
     GET_METRIC(tiflash_storage_page_write_duration_seconds, type_latch).Observe(watch.elapsedSeconds());
     watch.restart();
 
+    writers.push_back(&w);
+    w.cv.wait(apply_lock, [&] { return w.done || &w == writers.front(); });
+    GET_METRIC(tiflash_storage_page_write_duration_seconds, type_wait_in_group).Observe(watch.elapsedSeconds());
+    watch.restart();
+    if (w.done)
+        return {};
```
Comment (JaySon-Huang) [inline dbms/src/Storages/Page/V3/PageDirectory.cpp:1493]:
```suggestion
    if (w.done)
    {
        // the `applied_data_files` will be returned by the write
        // group owner, others just return an empty set.
        return {};
    }
```

**Review by JaySon-Huang — APPROVED** [review]
LGTM

**Comment by lidezhu** [comment]
The code for handling exception is not ready.

File: dbms/src/Storages/Page/V3/PageDirectory.cpp (hunk 2 above)
Comment (lidezhu) [reply dbms/src/Storages/Page/V3/PageDirectory.cpp]: Added.

**Comment by lidezhu** [comment]
@flowbehappy @CalvinNeo @JaySon-Huang Add some new code for handling exception, PTAL again

**Pull request merged**